### PR TITLE
Best Practices

### PR DESCRIPTION
**Pull request by Zeroshi**
Converted and rewrote some of the Best Practices from the presentation slide. The tutorials will need to be recreated ass well sincxe the dont exist any longer and would likely be outdated.

### CODE REVIEW

**Comment by SebastianStehle** [comment]
Cool stuff...What is still unclear for me is the best practices around the number of grains? If you are talking about 100 million IOT devices, is this a number that is acceptable for grains or do we need a grain per device "group"?

**Comment by Zeroshi** [comment]
How would you set up a device group? Are you saying something like thermostat model x group versus thermostat grain?

File: src/Documentation/resources/BestPractices [hunk 1]
```diff
@@ -0,0 +1,422 @@
+# Best Practices
```
Comment (ReubenBond) [inline src/Documentation/resources/BestPractices]: I think this file should be called best_practices.md, or at least end in .md

**Comment by Zeroshi** [comment]
Oh good call. I'll kill it and recommit

Sent from my Verizon, Samsung Galaxy smartphone
Get Outlook for Android<https://aka.ms/ghei36>
________________________________

File: src/Documentation/resources/BestPractices (hunk 1 above)
Comment (Zeroshi) [reply src/Documentation/resources/BestPractices]: Good call!

Comment (Zeroshi) [reply src/Documentation/resources/BestPractices]: done.

**Comment by Zeroshi** [comment]
> 
> 
> Cool stuff...What is still unclear for me is the best practices around the number of grains? If you are talking about 100 million IOT devices, is this a number that is acceptable for grains or do we need a grain per device "group"?

File: src/Documentation/resources/Best_Practices.md [hunk 2]
```diff
@@ -0,0 +1,422 @@
...
+    (TAP))
+
+-   [await](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/async/)
+    is the best syntax to use when composing async operations
+
+-   Common Scenarios:
+
+    -   Return a concrete value:
+
+        -   return Task.FromResult(value);
```
Comment (erictuvesson) [inline src/Documentation/resources/Best_Practices.md]: I think that code should always be kept in code blocks, it makes it easier to read and follow along. 😃

**Comment by SebastianStehle** [comment]
> How would you set up a device group? Are you saying something like thermostat model x group versus thermostat grain?

Either by something from the domain model, e.g. grain per customer, room, region, building...or just a simple module operation to create partitions.

File: src/Documentation/resources/Best_Practices.md [hunk 3]
```diff
@@ -0,0 +1,422 @@
...
+
+-   Failures can happen at any time and are a normal occurrence
+
+    -   Lost grains are automatically reactivated
+
+    -   Grains that are making in-process calls can fail or timeout
+
+    -   Orleans provides it best effort for grain delivery
+
+    -   Any network message can be lost and should be retired by the application
```
Comment (JanEggers) [inline src/Documentation/resources/Best_Practices.md]: retired  => retried

File: src/Documentation/resources/Best_Practices.md (hunk 2 above)
Comment (Zeroshi) [reply src/Documentation/resources/Best_Practices.md]: Ah yes! done! thanks!

File: src/Documentation/resources/Best_Practices.md (hunk 3 above)
Comment (Zeroshi) [reply src/Documentation/resources/Best_Practices.md]: thank you! corrected!

**Comment by Zeroshi** [comment]
> 
> 
> > How would you set up a device group? Are you saying something like thermostat model x group versus thermostat grain?
> 
> Either by something from the domain model, e.g. grain per customer, room, region, building...or just a simple modulo operation to create partitions.

Ah got it. This is based on the grain size in the document. 
"Grain Size
•	Optimal throughput is achieved by using multiple smaller grains rather than a few larger grains. However, the best practice of choosing a grain size and type is base on the application domain model.
–	Example: Users, Orders, etc."

If this doesn't answer your question, please let me know.

File: src/Documentation/resources/Best_Practices.md [hunk 4]
```diff
@@ -0,0 +1,422 @@
+# Best Practices
+
+Orleans was designed in 2014 with the intention of removing low-level
```
Comment (sergeybykov) [inline src/Documentation/resources/Best_Practices.md]: Orleans has been designed and built much earlier. It was running in production in late 2011. Also, the intent had nothing to do with improving the actor model per se. I think it's better to rephrase the into to something like the following.

"Orleans was built with the goal to greatly simplify building of distributed scalable applications, especially for the cloud. Orleans invented the [Virtual Actor Model](https://www.microsoft.com/en-us/research/wp-content/uploads/2016/02/Orleans-MSR-TR-2014-41.pdf) as an evolution of the Actor Model optimized for the cloud scenarios.

Grains (virtual actors) are the base building blocks of an Orleans-based application. They encapsulate state and behavior of application entities and maintain their lifecycle, The programming model of Orleans and the characteristics of its runtime fit some types of applications better than others. This document is intended to capture some of the tried and proven application patterns that work well in Orleans."

File: src/Documentation/resources/Best_Practices.md [hunk 5]
```diff
@@ -0,0 +1,422 @@
...
+abstractions from actor model frameworks. Grains (actors stored in memory) are
+used to maintain the lifecycle among other features. Orleans is based on the
+actor model and fits some projects better that others.
+
+## Orleans should be considered when:
+
+-   Significant number (hundreds to millions) of loosely coupled entities
```
Comment (sergeybykov) [inline src/Documentation/resources/Best_Practices.md]: I'm not sure how best to reflect the fact that there are two number here really - total number of entities, which can easily be in the billions and trillions, and the subset of that total number that is activated at any point in time.

User profiles are a good example. It is easy to have a grain for every person on Earth within a small cluster, so long as only a subset of users are active at any point in time.

File: src/Documentation/resources/Best_Practices.md [hunk 6]
```diff
@@ -0,0 +1,422 @@
...
+
+    -   Example: Internet of Things (IoT) – individual devices turning on/off
```
Comment (sergeybykov) [inline src/Documentation/resources/Best_Practices.md]: I would add some other examples: user profiles, purchase orders, application/game sessions, stocks.

File: src/Documentation/resources/Best_Practices.md [hunk 7]
```diff
@@ -0,0 +1,422 @@
...
+
+-   Entities are small enough to be single-threaded
+
+    -   Example: Determine if stock should be purchased based on current price
+
+-   Workload is interactive
+
+    -   Example: request-response, start/monitor/complete
+
+-   More than one server is expected or may be required\\
```
Comment (sergeybykov) [inline src/Documentation/resources/Best_Practices.md]: Typo - "\\"?

File: src/Documentation/resources/Best_Practices.md [hunk 8]
```diff
@@ -0,0 +1,422 @@
...
+
+    -   Orleans runs on a cluster which is expanded by adding servers to expand
+        the cluster
+
+-   Global coordination is not needed or on a smaller scale between a few
+    entities at a time
+
+    -   Orleans maintains its clusters and talks to other clusters. Each cluster
```
Comment (sergeybykov) [inline src/Documentation/resources/Best_Practices.md]: I don't see how this illustrates the topic. Maybe something along the lines of:
"Scalability and performance of execution is achieved by parallelizing and distributed a large number of mostly independent tasks with no single point of synchronization." instead?

File: src/Documentation/resources/Best_Practices.md [hunk 9]
```diff
@@ -0,0 +1,422 @@
...
+        is able to communicate through tcp/ip
+
+-   *Various entities are used at different times*
```
Comment (sergeybykov) [inline src/Documentation/resources/Best_Practices.md]: This is debatable. it is entirely possible to keep all entities of certain type in memory, for example, all users or all devices. And for some applications that's the best. Of course, the number of entities must be reasonable in those case. But in general, there's no requirement for the set of active entities to differ over time, even though that's a popular case.

File: src/Documentation/resources/Best_Practices.md [hunk 10]
```diff
@@ -0,0 +1,422 @@
...
+    -   Each grain maintains its own states and should not be shared.
+
+-   A small number of large entities that may be multithreaded
+
+    -   A microservice may be a better option when supporting complex logic in a
+        single service
+
+-   Global coordination and/or consistency is needed
+
+    -   Orleans maintains the silos within its own framework
```
Comment (sergeybykov) [inline src/Documentation/resources/Best_Practices.md]: Here again the example doesn't seem to fit the topic. Maybe "Such global coordination would severely limit performance of an Orleans-based application."?

File: src/Documentation/resources/Best_Practices.md [hunk 11]
```diff
@@ -0,0 +1,422 @@
...
+
+    -   Batch jobs, Single Instruction Multiple Data (SIMD) tasks
+
+    -   This depends on the need of the application and may be a fit for Orleans
+
+## Grains
+
+**Overview**:
+
+-   Actors are not objects, however they are similar
```
Comment (sergeybykov) [inline src/Documentation/resources/Best_Practices.md]: As a general rule, we try to refrain from talking about actors. I think it is better to talk about grains instead.

I would also start with a positive statement. For example, "Grains resemble objects. However, they are distributed, virtual, and asynchronous."

File: src/Documentation/resources/Best_Practices.md [hunk 12]
```diff
@@ -0,0 +1,422 @@
...
+
+        -   `var tasks = new List\<Task\>();  
+            foreach(var grain in grains)  
+            { tasks.Add(grain.Foo()) }  
+            await Task.WhenAll(tasks);  
+            DoMoreWork();`
+
+**Implementation of Grains**:
+
+-   When to use a [StatelessWorker]
```
Comment (sergeybykov) [inline src/Documentation/resources/Best_Practices.md]: I would start with the first cardinal rule or grain implementation - never perform a thread-blocking operation within grain code. Examples: synchronously waiting for an IO operation or a web service call, taking a lock, running a tight loop waiting for a condition, etc. All operations other than local computations must be explicitly asynchronous.

File: src/Documentation/resources/Best_Practices.md (hunk 4 above)
Comment (Zeroshi) [reply src/Documentation/resources/Best_Practices.md]: Corrected.

File: src/Documentation/resources/Best_Practices.md (hunk 5 above)
Comment (Zeroshi) [reply src/Documentation/resources/Best_Practices.md]: Corrected

File: src/Documentation/resources/Best_Practices.md (hunk 6 above)
Comment (Zeroshi) [reply src/Documentation/resources/Best_Practices.md]: corrected

File: src/Documentation/resources/Best_Practices.md (hunk 7 above)
Comment (Zeroshi) [reply src/Documentation/resources/Best_Practices.md]: corrected

File: src/Documentation/resources/Best_Practices.md (hunk 8 above)
Comment (Zeroshi) [reply src/Documentation/resources/Best_Practices.md]: corrected

File: src/Documentation/resources/Best_Practices.md (hunk 9 above)
Comment (Zeroshi) [reply src/Documentation/resources/Best_Practices.md]: corrected

File: src/Documentation/resources/Best_Practices.md (hunk 10 above)
Comment (Zeroshi) [reply src/Documentation/resources/Best_Practices.md]: corrected

File: src/Documentation/resources/Best_Practices.md (hunk 11 above)
Comment (Zeroshi) [reply src/Documentation/resources/Best_Practices.md]: That's a much better way to put it!

Comment (Zeroshi) [reply src/Documentation/resources/Best_Practices.md]: corrected

File: src/Documentation/resources/Best_Practices.md (hunk 12 above)
Comment (Zeroshi) [reply src/Documentation/resources/Best_Practices.md]: corrected

**Comment by SebastianStehle** [comment]
> > > How would you set up a device group? Are you saying something like thermostat model x group versus thermostat grain?
> > 
> > 
> > Either by something from the domain model, e.g. grain per customer, room, region, building...or just a simple modulo operation to create partitions.
> 
> Ah got it. This is based on the grain size in the document.
> "Grain Size
> • Optimal throughput is achieved by using multiple smaller grains rather than a few larger grains. However, the best practice of choosing a grain size and type is base on the application domain model.
> – Example: Users, Orders, etc."
> 
> If this doesn't answer your question, please let me know.

This is a typical consultancy answer: It depends.. :D

**Comment by Zeroshi** [comment]
> 
> 
> > > > How would you set up a device group? Are you saying something like thermostat model x group versus thermostat grain?
> > > 
> > > 
> > > Either by something from the domain model, e.g. grain per customer, room, region, building...or just a simple modulo operation to create partitions.
> > 
> > 
> > Ah got it. This is based on the grain size in the document.
> > "Grain Size
> > • Optimal throughput is achieved by using multiple smaller grains rather than a few larger grains. However, the best practice of choosing a grain size and type is base on the application domain model.
> > – Example: Users, Orders, etc."
> > If this doesn't answer your question, please let me know.
> 
> This is a typical consultancy answer: It depends.. :D

Ha, well it does depend. If you want to highest throughput then super small grains would be desired. That can create the issue of design, implementation, maintenance, etc. The way I go about it is based on its domain role. For instance, i would make a profile grain that talks to a payer grain. Depending on the complexity, i may break the payer grain into specifics such as debit, mastercard, etc.

File: src/Documentation/resources/Best_Practices.md [hunk 13]
```diff
@@ -0,0 +1,421 @@
...
+
+    -   A **timer** is an alternative method to write updates periodically.
+
+        -   The timer allows the application to determine the amount of
+            “eventual consistency”/statelessness allowed.
+
+        -   Timing (immediate/none/minutes) can also be controlled as to when to
+            update.
+
+    -   Each grain class can only be associated with one storage provider.
```
Comment (sergeybykov) [inline src/Documentation/resources/Best_Practices.md]: This part wouldn't be needed for `IPersistentState<TState>` or can be rephrased for it instead.

File: src/Documentation/resources/Best_Practices.md [hunk 14]
```diff
@@ -0,0 +1,421 @@
...
+
+-   MemoryStorage (Data stored in memory without durable persistence) is used
+    *only* for debugging and unit testing.
+
+-   AzureTableStorage
+
+    -   Configure the Azure storage account information with an optional
+        DeleteStateOnClear (hard or soft deletions)
+
+    -   Orleans serializer efficiently stores binary data in one Azure table
```
Comment (sergeybykov) [inline src/Documentation/resources/Best_Practices.md]: Binary serializer is not recommended. JSON is a more flexible option.

File: src/Documentation/resources/Best_Practices.md [hunk 15]
```diff
@@ -0,0 +1,421 @@
...
+        cell
+
+    -   Data size limit == max size of the Azure column which is 64kb of binary
+        data
+
+    -   Community contributed code that extends the use of multiple table
+        columns which increases the overall maximum size to 1mb.
+
+-   SharedStorageProvider will write data to multiple underlying storage
```
Comment (sergeybykov) [inline src/Documentation/resources/Best_Practices.md]: We removed sharded storage provided a while back.

File: src/Documentation/resources/Best_Practices.md [hunk 16]
```diff
@@ -0,0 +1,421 @@
...
+
+How to deal with Storage Operation Failures
+
+-   Grains and storage providers can await storage operations and *retry*
+    failures as needed
+
+-   Unhandled failures will propagate back to the caller and will be seen by the
+    client as a broken promise
+
+-   Other than the initial read, there is not a concept that automatically
```
Comment (sergeybykov) [inline src/Documentation/resources/Best_Practices.md]: If grain lets an `InconsistentStateException` escape, it gets automatically deactivated.

File: src/Documentation/resources/Best_Practices.md [hunk 17]
```diff
@@ -0,0 +1,421 @@
...
+    Clear, ReadStateAsync())
+
+-   The storage behavior is typically configurable (Batch writing, Hard or Soft
+    Deletions, etc.) and defined by the storage provider
+
+## Cluster Management
+
+-   Orleans automatically manages clusters
+
+    -   Worker roles (silo) can fail and join at any time which is handled
```
Comment (sergeybykov) [inline src/Documentation/resources/Best_Practices.md]: I think talking about worker roles (a.k.a Classic Cloud Services) is confusing here. I think "node" is a more general term.

File: src/Documentation/resources/Best_Practices.md [hunk 18]
```diff
@@ -0,0 +1,421 @@
...
+        automatically
+
+    -   A silo instance table exists for diagnostics
```
Comment (sergeybykov) [inline src/Documentation/resources/Best_Practices.md]: The tables exist first and foremost for the clustering protocol to work. It is also very helpful for diagnostics because it keeps history of all silos in the cluster.

File: src/Documentation/resources/Best_Practices.md [hunk 19]
```diff
@@ -0,0 +1,421 @@
...
+
+    -   There are also configuration options of an aggressive or a more lenient
+        failure detection
+
+-   Failures can happen at any time and are a normal occurrence
+
+    -   Lost grains are automatically reactivated
```
Comment (sergeybykov) [inline src/Documentation/resources/Best_Practices.md]: I think this may create a misunderstanding that grain can simply disappear. It is probably better to be more explicit. For example, "In case of a silo failure, grains there were activated on that silo get automatically reactivated later on other silos in the cluster."

File: src/Documentation/resources/Best_Practices.md [hunk 20]
```diff
@@ -0,0 +1,421 @@
...
+
+    -   Grains that are making in-process calls can fail or timeout
```
Comment (sergeybykov) [inline src/Documentation/resources/Best_Practices.md]: I'm not sure I understand what this is referring to.

File: src/Documentation/resources/Best_Practices.md [hunk 21]
```diff
@@ -0,0 +1,421 @@
...
+
+    -   Orleans provides it best effort for grain delivery
```
Comment (sergeybykov) [inline src/Documentation/resources/Best_Practices.md]: I would rephrase this as "Message delivery guarantee in Orleans is at-most-once", if that's what you were referring to.

File: src/Documentation/resources/Best_Practices.md [hunk 22]
```diff
@@ -0,0 +1,421 @@
...
+
+    -   All of the network messages can be lost and should be retried by the application
```
Comment (sergeybykov) [inline src/Documentation/resources/Best_Practices.md]: I would rephrase this as "It is a responsibility of the caller to retry any failed calls if needed."

File: src/Documentation/resources/Best_Practices.md [hunk 23]
```diff
@@ -0,0 +1,421 @@
...
+        if it is crucial
+
+        -   Common practice is to retry from end-to-end from the client/front
+            end
+
+-   Currently there is not a graceful shutdown
```
Comment (sergeybykov) [inline src/Documentation/resources/Best_Practices.md]: Graceful shutdown is supported.

File: src/Documentation/resources/Best_Practices.md [hunk 24]
```diff
@@ -0,0 +1,421 @@
...
+
+    -   A Azure upgrade or reboot is treated as a node failure
```
Comment (sergeybykov) [inline src/Documentation/resources/Best_Practices.md]: I would take this one out as it is confusing.

File: src/Documentation/resources/Best_Practices.md [hunk 25]
```diff
@@ -0,0 +1,421 @@
...
+
+## Deployment and Production Management
+
+Service Monitoring
+
+-   Orleans provided information
```
Comment (sergeybykov) [inline src/Documentation/resources/Best_Practices.md]: This is an outdated list.

File: src/Documentation/resources/Best_Practices.md [hunk 26]
```diff
@@ -0,0 +1,421 @@
...
+
+-   Create a custom performance counters
+
+Scaling out and in
+
+-   Monitor the Service-Level Agreement (SLA)
+
+-   Add or Remove instances
+
+-   Orleans automatically rebalances and takes advantage of the new hardware
```
Comment (sergeybykov) [inline src/Documentation/resources/Best_Practices.md]: This is not 100% accurate. We don't rebalance activated grains when a new silo is added to the cluster.

File: src/Documentation/resources/Best_Practices.md [hunk 27]
```diff
@@ -0,0 +1,421 @@
...
+
+## Logging and Testing
+
+-   Logging, Tracing, and Monitoring
+
+    -   GrainBase.GetLogger() exposes Info(), Warn(), Error(), Verbose()
+
+    -   .NET tracing and Runtimes trace to a local file by default
```
Comment (sergeybykov) [inline src/Documentation/resources/Best_Practices.md]: This is not true anymore.

File: src/Documentation/resources/Best_Practices.md [hunk 28]
```diff
@@ -0,0 +1,421 @@
...
+
+-   Add or Remove instances
+
+-   Orleans automatically rebalances and takes advantage of the new hardware
+
+## Logging and Testing
+
+-   Logging, Tracing, and Monitoring
+
+    -   GrainBase.GetLogger() exposes Info(), Warn(), Error(), Verbose()
```
Comment (sergeybykov) [inline src/Documentation/resources/Best_Practices.md]: The recommended way is to inject logger via DI. For example, like here - https://github.com/dotnet/orleans/blob/master/Samples/3.0/HelloWorld/src/HelloWorld.Grains/HelloGrain.cs#L14.

File: src/Documentation/resources/Best_Practices.md [hunk 29]
```diff
@@ -0,0 +1,421 @@
...
+
+    -   .NET tracing and Runtimes trace to a local file by default
+
+    -   Azure Diagnostics is also easily consumed
```
Comment (sergeybykov) [inline src/Documentation/resources/Best_Practices.md]: I would take this out.

File: src/Documentation/resources/Best_Practices.md [hunk 30]
```diff
@@ -0,0 +1,421 @@
...
+
+    -   Logger.LogConsumer.Add() allows the developer to choose their own
```
Comment (sergeybykov) [inline src/Documentation/resources/Best_Practices.md]: This is outdated since we switched to leveraging `Microsoft.Extensions.Logging` in 2.0.

File: src/Documentation/resources/Best_Practices.md [hunk 31]
```diff
@@ -0,0 +1,421 @@
...
+        logging consumer
+
+    -   Default tracing levels can be overridden for grains and runtime
+        components
+
+        -   Example: `<TraceLevelOverrideLogPrefix="Application"
```
Comment (sergeybykov) [inline src/Documentation/resources/Best_Practices.md]: This is all gone since 2.0.

File: src/Documentation/resources/Best_Practices.md [hunk 32]
```diff
@@ -0,0 +1,421 @@
...
+            TraceLevel=“Verbose">`
+
+        -   Example: `<TraceLevelOverrideLogPrefix=“Runtime"
+            TraceLevel=“Warning">`
+
+Testing
+
+-   UnitTestingBase class is used for unit testing
```
Comment (sergeybykov) [inline src/Documentation/resources/Best_Practices.md]: `UnitTestingBase` is gone.

File: src/Documentation/resources/Best_Practices.md [hunk 33]
```diff
@@ -0,0 +1,421 @@
...
+
+-   Starting two or more silos in app domains and a client in the main app
```
Comment (sergeybykov) [inline src/Documentation/resources/Best_Practices.md]: This is outdated since 2.0.

File: src/Documentation/resources/Best_Practices.md [hunk 34]
```diff
@@ -0,0 +1,421 @@
...
+    domain
+
+-   Simplifies the versions of what is used internally
+
+Troubleshooting
+
+-   Logs (Silo and Frontend logging)
+
+    -   WAD may not pick up the logging in the event of a startup failure
```
Comment (sergeybykov) [inline src/Documentation/resources/Best_Practices.md]: This is Azure specific and outdated. I would take it out.

File: src/Documentation/resources/Best_Practices.md [hunk 35]
```diff
@@ -0,0 +1,421 @@
...
+
+    -   Remote Desktop Protocol (RDP) to the machines for verification
```
Comment (sergeybykov) [inline src/Documentation/resources/Best_Practices.md]: This is not the recommended way anymore.

File: src/Documentation/resources/Best_Practices.md [hunk 36]
```diff
@@ -0,0 +1,421 @@
...
+-   Silo isn’t starting
+
+    -   Check OrleansSiloInstances to determine if the silo registered there.
+
+    -   Make sure that firewall is open for TCP ports: 11111 and 30000
+
+    -   Check the logs and don’t forget that there is an extra log for startup
+        errors
+
+    -   RDP into the server or worker role instance and attempt to start it
```
Comment (sergeybykov) [inline src/Documentation/resources/Best_Practices.md]: I would take this out.

File: src/Documentation/resources/Best_Practices.md [hunk 37]
```diff
@@ -0,0 +1,421 @@
+# Best Practices
+
+Orleans was built with the goal to greatly simplify building of distributed scalable applications, especially for the cloud. Orleans invented the Virtual Actor Model as an evolution of the Actor Model optimized for the cloud scenarios.
+
+Grains (virtual actors) are the base building blocks of an Orleans-based application. They encapsulate state and behavior of application entities and maintain their lifecycle, The programming model of Orleans and the characteristics of its runtime fit some types of applications better than others. This document is intended to capture some of the tried and proven application patterns that work well in Orleans.
```
[the remark this answers is not in the document]
Comment (Zeroshi) [reply src/Documentation/resources/Best_Practices.md]: Thank you. That does make it much easier to maintain.

File: src/Documentation/resources/Best_Practices.md [hunk 38]
```diff
@@ -0,0 +1,421 @@
...
+
+## Orleans should be considered when:
+
+-   Significant number (hundreds, millions, billions, and even trillions) of loosely coupled entities. To put the number in perspective, Orleans can easily create a grain for every person on Earth in a small cluster, so long as a subset of that total number is active at any point in time. 
+
+    -   Example: Additonal examples are: user profiles, purchase orders, application/game sessions, stocks
```
[the remark this answers is not in the document]
Comment (Zeroshi) [reply src/Documentation/resources/Best_Practices.md]: that would be due to a lack of sleep. thank you for looking over the document

File: src/Documentation/resources/Best_Practices.md (hunk 20 above)
Comment (Zeroshi) [reply src/Documentation/resources/Best_Practices.md]: This was based on the best practices presentation. I was assuming that it meant grains may fail or timeout when making calls. I re-wrote it as, "Grains have an ability to timeout. A retry solution such as Polly can assist with retries."

File: src/Documentation/resources/Best_Practices.md (hunk 23 above)
Comment (Zeroshi) [reply src/Documentation/resources/Best_Practices.md]: Thank you. I asked that one in Gitter because I was pretty sure if was. Thank you for validating.

File: src/Documentation/resources/Best_Practices.md (hunk 25 above)
Comment (Zeroshi) [reply src/Documentation/resources/Best_Practices.md]: removed for now

File: src/Documentation/resources/Best_Practices.md (hunk 26 above)
Comment (Zeroshi) [reply src/Documentation/resources/Best_Practices.md]: That's very insightful!

File: src/Documentation/resources/Best_Practices.md (hunk 27 above)
Comment (Zeroshi) [reply src/Documentation/resources/Best_Practices.md]: Does Dynatrace work?

Comment (Zeroshi) [reply src/Documentation/resources/Best_Practices.md]: removed the .net tracing for the moment

File: src/Documentation/resources/Best_Practices.md [hunk 39]
```diff
@@ -0,0 +1,421 @@
...
+extensible storage functionality.
+
+-   Tutorial: *Needs to be created*
+
+**Overview**:
+
+-   Orleans.IGrainState is extended by a .NET interface which contains fields
+    that should be included in the grain’s persisted state.
+
+-   GrainBase\<T\> is extended by the grain class that adds a strongly typed
```
[the remark this answers is not in the document]
Comment (Zeroshi) [reply src/Documentation/resources/Best_Practices.md]: Thanks for the catch!

File: src/Documentation/resources/Best_Practices.md (hunk 13 above)
Comment (Zeroshi) [reply src/Documentation/resources/Best_Practices.md]: Updated to, "PersistetState classes, like other grain classes, can only be associated with one storage provider." Thanks again!

**Comment by sergeybykov** [comment]
Thank you, @Zeroshi!
